src/API/Google/AdsCampaign.php

```diff
@@ -176,21 +185,35 @@ public function get_campaign( int $id ): array {
 	 */
 	public function create_campaign( array $params ): array {
 		try {
+			$base_country = $this->container->get( WC::class )->get_base_country();
+			$location_ids = array_map(
+				function ( $country_code ) {
+					return $this->google_helper->find_country_id_by_code( $country_code );
```

Comment:
Thanks for catching this critical error! I updated in e790283 but not 100% satisfy with the solution, because there will be an unused `$this->google_helper` inside `CampaignController`. Would like to know if there's any better solution when a trait needs to call a service's method?

Reply:
I agree that it's probably not the best pattern to use. GoogleHelper used to be a trait as well, and for a similar reason it got converted to a Service. Also the way you are doing it now means if any controller class has google_helper it will do a stricter validation, otherwise not. Most likely the stricter check is useful for other controllers as well, but we'd have to check that in each controller.
My suggestion was more along the lines of swapping out `get_country_code_validate_callback` to a local function within the controller which uses both the function in the trait and the google_helper function.

Another option if we want to make the validation reusable would be to add a GoogleHelperAwareInterface (and related trait) like is done for the ISO service. I'd still create a separate function like `get_supported_country_code_validate_callback` so the controller can decide which one to use.

Reply:
Another thought was, since there are no other controllers that enforce only supported countries then we could decide to follow that same pattern. The UI already has a list of supported countries to work with, so generally it won't send an API request with unsupported countries.

Although we would still need to fix AdsCampaign to make sure it filters out unsupported countries (or null values when converting the ID which is equal to an unsupported country). That would be a much simpler approach. Yes we would have a less strict controller, but as long as it's not possible to produce a fatal it is a similar pattern we have for other controllers.

Reply:
I finished implementing `GoogleHelperAwareInterface` via 0ffa8dd before seeing your last comment.. 😅 I think if you think the `GoogleHelperAwareInterface` implementation is OK we can go with this solution, but I'll still create a followup issue to make sure other controllers to enforce supported countries if they need to.

Also, I added an `array_filter` in `AdsCampaign.php` via 1f139d6 to make sure we don't pass location IDs array with null value to `AdsCampaignCriterion`'s `create_operation` method.

Thanks for all the thorough thoughts! 🙏